core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSource.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.exchange;
+
+import io.airlift.slice.Slice;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.Closeable;
+import java.util.concurrent.CompletableFuture;
+
+@ThreadSafe
+public interface ExchangeSource
+        extends Closeable
+{
+    CompletableFuture<?> NOT_BLOCKED = CompletableFuture.completedFuture(null);
+
+    /**
+     * Returns a future that will be completed when the exchange source becomes
+     * unblocked.  If the exchange source is not blocked, this method should return
+     * {@code NOT_BLOCKED}
+     */
+    CompletableFuture<?> isBlocked();
+
+    /**
+     * Will this exchange source product more data?
```

Comment:
Document in terms of the guarantees it provides: "once isFinished returns true, read() will never return a non-null result"